connectors/sources/jira.py

```diff
@@ -0,0 +1,467 @@
+                yield {
+                    "_id": f"{project['name']}-{project['id']}",
+                    "_timestamp": iso_utc(),
+                    "Type": "Project",
+                    "Project": project,
+                }
+
+    async def _get_issues(self):
+        """Get issues with the help of REST APIs
+
+        Yields:
+            Dictionary: Jira issue to get indexed
+            issue (dict): Issue response to fetch the attachments
+        """
+        async for response in self._api_call(
+            url_name=ISSUES, is_pagination=True, maxResults=FETCH_SIZE, startAt=0
+        ):
+            for issue in response.get("issues", []):
+                async for response in self._api_call(
+                    url_name=ISSUE_DATA, id=issue["key"]
+                ):
+                    issue = await response.json()
+                    if issue:
+                        response_fields = issue.get("fields")
+                        yield {
+                            "_id": f"{response_fields['project']['name']}-{issue['key']}",
+                            "_timestamp": response_fields["updated"],
+                            "Type": response_fields["issuetype"]["name"],
+                            "Issue": response_fields,
+                        }, issue
+
+    async def _get_attachments(self, attachments, issue_key):
+        """Get attachments of a specific issue
+
+        Args:
+            attachments (list): List of attachments for an issue
+            issue_key (str): Issue key for generating `_id` field
+
+        Yields:
+            Dictionary: Jira attachment on the given issue
+            attachment (dict): Attachment dictionary for extracting the content
+        """
+        for attachment in attachments:
+            yield {
+                "_id": f"{issue_key}-{attachment['id']}",
+                "title": attachment["filename"],
+                "type": "Attachment",
+                "issue": issue_key,
+                "_timestamp": attachment["created"],
+                "size": attachment["size"],
+            }, attachment
+
+    async def _grab_content(self, attachments, issue_key):
+        """Coroutine to add attachments to Queue and get content
+
+        Args:
+            attachments (list): List of attachments for an issue
+            issue_key (str): Issue key for generating `_id` field
+        """
+        async for content, attachment in self._get_attachments(
+            attachments=attachments, issue_key=issue_key
+        ):
+            if attachment:
+                await self.queue.put(
+                    (
+                        content,
+                        partial(
+                            self.get_content,
+                            issue_key=issue_key,
+                            attachment=copy(attachment),
+                        ),
+                    )
+                )
+
+    async def get_docs(self, filtering=None):
```

Comment:
Something feels weird in this function - framework is able to handle getting attachments from 3rd-party system, all you need to do in your `get_docs` is call `yield document, partial(self.get_content, document)` and framework will call the partial function you provided to download the document and update the document content with extracted content.

Why did you do it differently here?

Reply:
This is a way that was suggested by Tarek to achieve the concurrency at the connector level. You can refer this Tarek's comment - https://github.com/elastic/connectors-python/issues/256#issuecomment-1367643403